selftests/unit/test_utils_filelock.py

```diff
@@ -0,0 +1,40 @@
+import os
+import shutil
+import tempfile
+import unittest
+
+from avocado.utils.filelock import AlreadyLocked
+from avocado.utils.filelock import FileLock
+
+
+class TestFileLock(unittest.TestCase):
+
+    def setUp(self):
+        self.basedir = tempfile.mkdtemp(prefix='avocado_' + __name__)
+        self.filename = os.path.join(self.basedir, 'file.img')
+        self.content = 'Foo bar'
+        with open(self.filename, 'w') as f:
+            f.write(self.content)
+
+    def _readfile(self):
+        with FileLock(self.filename):
+            with open(self.filename, 'r') as f:
+                return f.read()
+
+    def test_readfile(self):
+        self.assertEqual(self._readfile(), self.content)
+
+    def test_locked_by_me(self):
+        with FileLock(self.filename):
+            self.assertRaises(AlreadyLocked, self._readfile)
+
+    def test_locked_by_other(self):
+        with open(self.filename+'.lock', 'w') as f:
+            f.write('1')
+        self.assertRaises(AlreadyLocked, self._readfile)
+
+    def tearDown(self):
+        shutil.rmtree(self.basedir)
+
+if __name__ == "__main__":
+    unittest.main()
```

Comment:
@apahim I wrote the following script to test the `filelock` module:

```python
import os
import time
import random
import multiprocessing

from avocado.utils.filelock import FileLock


def action(args):
    path, members = args
    max_individual_timeout = 1
    max_timeout = max_individual_timeout * members
    print('%s will try to acquire lock' % os.getpid())
    with FileLock(path, max_timeout) as lock:
        sleeptime = random.random()
        print('%s holding lock %s for %2.2fs' % (os.getpid(),
                                                 lock.filename,
                                                 sleeptime))
        time.sleep(sleeptime)
    print('%s released lock' % os.getpid())

if __name__ == '__main__':
    # configure the script here
    members = 100
    path = '/tmp/foo'

    # run in parallel
    pool = multiprocessing.Pool(members)
    args = [(path, members)] * members
    pool.map(action, args)

```

With small values for `members`, say `4`, it works really great.  For larger numbers, say with `100`, it shows some race conditions pretty much on every execution.  Errors include:

```
Avocado.utils.filelock.AlreadyLocked: File is locked by someone else.
```

Which seems to be caused by trying to create the lock file, failing it, and then trying to read it while it is not yet written.  Adding `os.O_SYNC` to flags *seems* to improve this situtation a bit, but it's still not failproof.

I also got `IOErrors`, which looks like come from races between `os.open(self.filename, flags)`, and `open(self.filename, 'r')`, given that this second statement could fail if the lock file doesn't exist anymore (lock was released by other process).